File: python/quantum/test/test_algorithms.py
# test_algorithms.py

import sys
sys.path.append('../..')

import unittest
from math import pi, sqrt
import numpy
from numpy import array, ones
from numpy.fft import fft

import quantum
from quantum.algorithms import grover_invert, grover_search, qft, shor_factor
from quantum.gate_factory import hadamard_gate, phase_shift_gate
from quantum.quantum_gate import tensor_power
from quantum.qubit_system import QubitSystem
from quantum._util import EPS

class Test_algorithms(unittest.TestCase):
    # Test quantum Fourier transform
    def test_qft(self):
        N = 4
        q = QubitSystem(N)
        # put q into a non-uniform mixed state
        hadamard_gate(N) * q
        tensor_power(phase_shift_gate(pi), N) * q

        # classical FFT of coefficients, normalized to be unitary like the QFT
        yhat = fft(q._QubitSystem__coeffs) / sqrt(2**N)
        qft(q)
        self.assertTrue(numpy.max(numpy.abs(q._QubitSystem__coeffs - yhat)) \
                        < EPS)


    # Tests Grover's search algorithm
    def test_grover_search(self):
        # List of strings to search
        LIST = ['John', 'Paul', 'George', 'Ringo', 'Paul']

        # Test that it can correctly find each string in the list
        for item in LIST:
            self.assertEqual(LIST[grover_search(item, LIST)], item)
        # Test that it returns -1 when the item is missing
        self.assertEqual(grover_search('Yoko', LIST), -1)


    # Test Grover's function inversion algorithm
    def test_grover_inversion(self):
        # function for testing: inverse of discrete log problem
        f = lambda x: pow(13, x, 8) if x >= 0 else None
        n = 3 # number of bits to allow as input to f

        # Test that it can solve the problem for each possible y value
        yrange = map(f, range(2**n))
        for y in yrange:
            self.assertEqual(f(grover_invert(f, y, n)), y)
        # Test that it returns -1 when the value is not in the range
        self.assertEqual(grover_invert(f, 3, n), -1)


    # Test Shor's algorithm
    def test_shor(self):
        composites = [4, 6, 9, 10, 12]
        factors = map(shor_factor, composites)
        # test that the factors are indeed valid factors
        self.assertEqual([a % b for a, b in zip(composites, factors)],
                         [0] * len(composites))
        # test that the factors are non-trivial
        self.assertTrue(all(array(factors) != ones(len(composites))))
        self.assertTrue(all(array(factors) != array(composites)))

        # Test that it returns 1 for prime numbers
        primes = [2, 3, 7, 29]
        self.assertEqual(map(shor_factor, primes), [1] * len(primes))



if __name__ == '__main__':
    unittest.main()

File: cpp/quantum/test/test_algorithms.cpp
#include "algorithms.h" 
#include <gtest/gtest.h>
using std::string;
using namespace quantum_algorithm_simulator; 

// Tests Grover's search algorithm
TEST (AlgorithmsTest, TestGroverSearch) {
	// List of strings to search 
	static const string LIST[] = {"John", "Paul", "George", "Ringo", "Paul"}; 
	static const int LIST_LENGTH = 5; 
	int index; 
	
	// Test that it can correctly find each string in the list
	for (int i = 0; i < LIST_LENGTH; i++) { 
		index = grover_search(LIST[i], LIST, LIST_LENGTH); 
		EXPECT_GE(index, 0); 
		EXPECT_LT(index, LIST_LENGTH); 
		if (0 <= index && index < LIST_LENGTH) { 
			EXPECT_EQ(LIST[i], LIST[index]); 
		}
	}

	// Test that it returns -1 when the item is missing
	index = grover_search("Yoko", LIST, LIST_LENGTH); 
	EXPECT_EQ(-1, index); 
}



// Function to use to test Grover's algorithm: inverse of discrete log problem
static int f (int x) { return ((int) pow(13, x)) % 8; }

// Tests Grover's function inversion algorithm
TEST (AlgorithmsTest, TestGroverInversion) {
	static const int N = 3; // number of bits to allow as input to f
	int x; 
	
	// Test that it can correctly solve the problem for each possible y value
	for (int i = 0; i < pow(2, N); i++) { 
		x = grover_invert(f, f(i), N); 
		EXPECT_GE(x, 0); 
		if (x >= 0) {
			EXPECT_EQ(f(i), f(x)); 
		}
	}

	// Test that it returns -1 when the item is missing
	x = grover_invert(f, 3, N); 
	EXPECT_EQ(-1, x); 
}


// Test Shor's algorithm
TEST (AlgorithmsTest, TestShor) {
	int composites[] = {4, 6, 9, 10, 12}; 
	const int LENGTH = 5; 

	for (int i = 0; i < LENGTH; i++) {
		int n = composites[i]; 
		int p = shor_factor(n); 
		// test that p is a factor
		EXPECT_EQ(0, n % p); 
		// test that p is not trivial
		EXPECT_NE(1, p); 
		EXPECT_NE(n, p); 
	}

	// Test that it returns 1 for prime numbers
	int primes[] = {2, 3, 7, 29}; 
	const int PRIMES_LENGTH = 4; 
	for (int i = 0; i < PRIMES_LENGTH; i++) { 
		int p = primes[i]; 
		EXPECT_EQ(1, shor_factor(p)); 
	}
}






